상품 CRUD 리팩터링

JunBe opened the pull request:
## #️⃣ Issue Number

#29 

## 📝 작업 설명

코드 가독성 향상을 위한 리팩터링

## 📸스크린샷 (선택)   

## 💬 리뷰 요구사항

필요 없을 것 같은 라인들이나 더 좋은 코드 구현 방식이 있다면 알려주시면 감사하겠습니다!

## ✅ PR Checklist

- [X] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [X] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

JunBe committed:
refactor: 조회 시 기본 정렬 기준을 최신 순으로 변경

- 기존 Default 값이 asc로 설정되어 오래된 순으로 출력되던 문제를 수정
- Default 값을 desc로 변경하여 최신 순으로 반환되도록 수정

#Resolves: 29

JunBe committed:
refactor: 등록 시간을 상대적으로 표시하도록 수정

- 등록 시간을 기준으로 초, 분, 시간, 일, 주, 개월, 년 단위로 상대적인 시간 표시
- "n초 전", "n분 전" 등의 형식으로 나타나도록 변경

#Resolves: 29

JunBe committed:
feat: 상품 상세 조회 API 추가

- /api/products/{productId} GET 요청을 통해 특정 ID 상품의 정보 조회 기능 추가

#Resolves: 29

JunBe committed:
refactor: DTO 변환 함수 Builder 패턴으로 변경

- Setter를 이용하는 방식에서 Builder 패턴으로 변경하여 가독성 향상

#Resolves: 29

JunBe committed:
refactor: 상대 시간 계산 함수를 TimeUtils로 분리

- 시간 계산 관련 기능들을 util 패키지의 TimeUtils 클래스로 이동

Resolves: #29

JunBe committed:
style: 불필요한 라인 삭제

Resolves: #29

ahyeonkong commented on src/main/java/com/market/saessag/domain/product/service/ProductService.java:
```diff
@@ -59,24 +60,28 @@ public Page<ProductResponse> searchProducts(int page, int size, String title, St
-        productDTO.setAddedDate(product.getAddedDate().toString());
-        productDTO.setStatus(product.getStatus().toString());
-        productDTO.setUser(userDTO);
+        return ProductResponse.builder()
+                .productId(product.getProductId())
+                .photo(product.getPhoto())
+                .title(product.getTitle())
+                .price(product.getPrice())
+                .description(product.getDescription())
+                .meetingPlace(product.getMeetingPlace())
+                .addedDate(TimeUtils.getRelativeTime(product.getAddedDate()))
+                .status(product.getStatus().toString())
+                .user(UserResponse.builder()
+                        .id(user.getId())
+                        .nickname(user.getNickname())
+                        .profileUrl(user.getProfileUrl())
+                        .build())
+                .build();
+    }
```
Controller와 Service 중 Service에서만 Builder를 사용하신 이유가 있으실까요?
구현할 수 있는 방법은 여러가지가 있는 것으로 알고 있습니다!

JunBe replied:
Service에서만 Builder 패턴을 사용한 이유는 코드가 너무 길어져 가독성을 높이기 위함이었습니다 
Controller는 서비스에서 처리된 로직을 단순히 호출하는 역할이기 때문에 Builder 패턴을 사용하면 오히려 가독성을 해칠 수 있다고 생각합니다!

JinTaekLim replied:
위의 코드는 Builder 내에서도 Builder를 사용하는 등의 이유로 가독성이 좋다! 라고 느껴지는 코드는 아닌 것같다는 생각이 드는데, 혹시 생각하신 다른 방안은 없으실까요?

JunBe replied:
Builder내에 Builder를 사용하는게 가독성이 떨어진다면 UserResponse.builder() 부분을 메서드로 추출해서 사용하는 것도 좋을 것 같다고 생각합니다!

ahyeonkong commented on src/main/java/com/market/saessag/domain/product/controller/ProductController.java:
```diff
@@ -23,11 +23,17 @@ public Page<ProductResponse> searchProducts(
+    }
+
     //상품 생성
     @PostMapping
```
HTTP 메서드의 일관된 응답 형식을 반환하기 위해 ApiResponse를 활용해보는 건 어떨까요?

ahyeonkong commented on src/main/java/com/market/saessag/domain/product/controller/ProductController.java:
```diff
@@ -23,11 +23,17 @@ public Page<ProductResponse> searchProducts(
         return productService.searchProducts(page, size, title, nickname, sort, direction);
     }
 
+    //상세 조회
+    @GetMapping("/{productId}")
+    public ProductResponse getProductDetail(@PathVariable Long productId) {
+        return productService.getProductDetail(productId);
+    }
+
```
일관된 응답 형식을 반환하기 위해 모든 HTTP 메서드에 ApiResponse를 활용해보는 건 어떨까요?

JunBe replied:
넵 추가해서 다시 커밋해보겠습니다!

ahyeonkong commented on src/main/java/com/market/saessag/domain/user/dto/UserResponse.java:
```diff
@@ -1,10 +1,11 @@
 
 @Getter
-@Setter
+@Builder
```
DTO의 필드 값이 변경될 일이 없다면 final 키워드를 사용하는 것이 어떨까요?

ahyeonkong commented on src/main/java/com/market/saessag/domain/user/dto/UserResponse.java:
```diff
@@ -1,10 +1,11 @@
-@Setter
+@Builder
 public class UserResponse {
     private Long id;
```
DTO의 필드 값이 변경될 일이 없다면 final 키워드를 사용하는 것이 어떨까요?
```suggestion
    private final Long id;
```

JunBe replied:
좋은 것 같습니다! 상품 DTO도 변경될 일 없는 필드는 final 키워드 추가 해 보겠습니다

ahyeonkong commented:
현재 모든 코드를 보면, 필드 주입 방식이랑 생성자 주입 방식이 혼용된 것 같아요! 
@Autowired를 이용한 필드 주입은 편리하지만, 단점이 많이 존재하기 때문에 생성자 주입 방식을 사용해보는 게 어떨까 하는 의견입니다. @RequiredArgsConstructor를 사용하거나 직접 생성자 주입 방식을 구현하는 방법이 있네요!


```
@RestController
@RequestMapping("/api/products")
public class ProductController {
    @Autowired
    private ProductService productService;
```
각각의 주입 방식을 비교해보시고, 의견 주세요!

JunBe committed:
refactor: ApiResponse를 사용하여 컨트롤러 응답 구조 개선

- 모든 컨트롤러 메서드에서 ApiResponse를 사용하여 일관된 응답 구조를 반환하도록 변경
- 상품 삭제 시 존재하지 않는 경우 404 응답 처리 추가

Resolves: #29

JunBe committed:
refactor: DTO 필드에 불변성을 보장하기 위해 final 키워드 추가

- User 엔티티의 id 필드, Product 엔티티의 productId, addedDate, user 필드는 한 번 결정된 이후로 수정이 필요 없으므로 final 키워드를 추가

Resolves: #29

JunBe committed:
refactor: 필드 주입 방식에서 생성자 주입 방식으로 변경

- 코드의 유지보수성 향상을 위해 생성자 주입 방식으로 변경

Resolves: #29

JunBe commented:
현재 필드 주입 방식으로 작성된 라인들을 살펴보니 한 번 생성된 이후에는 변경될 일이 없는 것 같습니다. 
따라서 말씀해 주신 대로 final 키워드를 이용해 @RequiredArgsConstructor를 사용한 생성자 주입으로 변경하면 테스트 작성 시 의존성 주입 측면이나, 코드의 유지보수성도 더욱 좋아질 것 같습니다!

JinTaekLim commented:
> 현재 모든 코드를 보면, 필드 주입 방식이랑 생성자 주입 방식이 혼용된 것 같아요! @Autowired를 이용한 필드 주입은 편리하지만, 단점이 많이 존재하기 때문에 생성자 주입 방식을 사용해보는 게 어떨까 하는 의견입니다. @requiredargsconstructor를 사용하거나 직접 생성자 주입 방식을 구현하는 방법이 있네요!
> 
> ```
> @RestController
> @RequestMapping("/api/products")
> public class ProductController {
>     @Autowired
>     private ProductService productService;
> ```
> 
> 각각의 주입 방식을 비교해보시고, 의견 주세요!

찾아보시고 두 방법의 자세한 장단점도 공유해주시면 좋을 것같아요!

JinTaekLim commented on src/main/java/com/market/saessag/domain/product/controller/ProductController.java:
```diff
@@ -3,48 +3,92 @@
     //통합 조회 (제목, 닉네임, 정렬기준)
     @GetMapping()
-    public Page<ProductResponse> searchProducts(
+    public ResponseEntity<ApiResponse<Page<ProductResponse>>> searchProducts(
```
현재는 ResponseEntity 내부에 ApiResponse 내부에 Page 내부에 ProductResponse 가 들어가있는 구조인데, 너무 복잡하다는 느낌이 드는 것같아요. 생각해두신 다른 방법은 없으실까요?

JunBe replied:
ApiResponse를 빼고 ResponseEntity만 사용하여 상태만 전달하거나, ResponseEntity를 빼고 ApiResponse만 사용하여 상태와 데이터 모두 전달하는 방법 생각해 봤는데 어떻게 생각하시나요??

ahyeonkong replied:
저는 ApiResponse만 사용하는 방법이 좋다고 생각합니다!
ApiResponse 클래스를 통해 성공/실패 여부, 데이터, 메시지를 포함하는 일관된 응답 구조를 만들 수 있습니다. HTTP 상태 코드는 @ResponseStatus 어노테이션이나 예외 처리를 통해 자동으로 처리할 수 있으므로, ResponseEntity를 굳이 사용할 필요가 없다고 생각합니다!
다른 의견 있으실까요?

JinTaekLim commented on src/main/java/com/market/saessag/domain/product/controller/ProductController.java:
```diff
@@ -3,48 +3,92 @@
             @RequestParam(defaultValue = "10") int size,
             @RequestParam(required = false) String title,
             @RequestParam(required = false) String nickname,
             @RequestParam(required = false) String sort, //정렬 기준 (price, addedDate 등)
```
다른 방법을 택하지 않고 RequestParm을 사용하신 이유가 있으실까요?

JunBe replied:
PathVariable 사용했을 때 닉네임이나 제목으로 검색하기 위해 일관된 경로가 아닌 `products/title/{제목}`, `products/nickname/{닉네임}` 와 같은 식으로 다른 경로로 요청을 보내야 한다는 점에서 API 구조가 복잡해 진다고 생각했습니다.
RequestParam을 사용하면 `products` 경로에서 파라미터를 이용하여 다양한 조건을 조회할 수 있기 때문에 사용했습니다!

JinTaekLim commented on src/main/java/com/market/saessag/domain/product/controller/ProductController.java:
```diff
@@ -3,48 +3,92 @@
                                  @RequestBody Product product) {
-        return productService.updateProduct(productId, product);
+        ProductResponse updatedProduct = productService.updateProduct(productId, product);
+        ApiResponse<ProductResponse> response = ApiResponse.<ProductResponse>builder()
```
ApiResponse 내부의 값을 매번 일일이 지정해줄 필요가 있을까요?
만약 있다고 한다면, 해당 값을 builder가 아니라 생성자로 정의하는게 조금 더 깔끔하지 않을까하는 생각입니다!

JunBe replied:
맞는 것 같습니다! 확실히 메서드마다 builder로 값을 넣어주니 길어지는 느낌이 있네요
생성자로 정의하면 더 깔끔해지고 보기 좋아질 것 같습니다!

JunBe committed:
chore: AWS S3 설정 추가 및 라이브러리 의존성 추가

- workflow.yml에 S3 버킷 설정 추가
- AWS SDK 의존성 추가
- 환경 변수를 이용한 엑세스 키, 시크릿 키 접근
- S3 활용을 위한 세팅

Resolves: #29

JunBe committed:
feat: S3에 사진 업로드, 조회 기능 추가

- 사진 관련 엔드포인트 허용

- /api/photos/upload 엔드포인트
  - POST 요청으로 사진 업로드 가능
  - 키 값은 files이며 여러 사진 업로드 가능

- /api/photos 엔드포인트
  - keys 파라미터로 반환 받은 key 값 사용
  - key에 대한 value는 프리사인 URL
  - 프리사인 URL로 사진 조회 가능

Resolves: #29

JunBe committed:
refactor: 등록 날짜 현재 기준으로 입력되도록 변경

- 기존에 직접 입력했던 등록 날짜를 현재 시간 기준으로 자동 등록되도록 변경

Resolves: #29

JunBe committed:
refactor: 상품 사진 필드 타입을 String에서 List로 변경

- 기존 String 타입에선 한 장의 사진 밖에 입력을 받을 수 있었음
- 여러 사진을 입력 받기 위해 List<String> 타입으로 변경

Resolves: #29

JunBe committed:
refactor: Product를 직접 받는 방식에서 ProductRequest로 변경

- Product 대신 ProductRequest를 요청으로 받도록 수정
- ProductRequest DTO에 필요한 필드 추가 및 getter 구현
- 서비스 계층에서 ProductRequest를 기반으로 Product 엔티티 생성

Resolves: #29

ahyeonkong reviewed (approved):
수고하셨습니다!
